src/sail_lean_backend/pretty_print_lean.ml

```diff
@@ -264,42 +273,73 @@ let string_of_exp_con (E_aux (e, _)) =
   | E_vector _ -> "E_vector"
   | E_let _ -> "E_let"
 
-let rec doc_exp ctx (E_aux (e, (l, annot)) as full_exp) =
+let wrap_with_pure (needs_return : bool) (d : document) =
+  if needs_return then parens (nest 2 (flow space [string "pure"; d])) else d
+
+let wrap_with_left_arrow (needs_return : bool) (d : document) =
+  if needs_return then parens (nest 2 (flow space [string "←"; d])) else d
+
+let rec doc_exp (as_monadic : bool) ctx (E_aux (e, (l, annot)) as full_exp) =
   let env = env_of_tannot annot in
+  let d_of_arg arg =
+    let arg_monadic = effectful (effect_of arg) in
+    wrap_with_left_arrow arg_monadic (doc_exp arg_monadic ctx arg)
+  in
+  let d_of_field (FE_aux (FE_fexp (field, e), _) as fexp) =
+    let field_monadic = effectful (effect_of e) in
+    doc_fexp field_monadic ctx fexp
+  in
   match e with
-  | E_id id -> string (string_of_id id) (* TODO replace by a translating via a binding map *)
-  | E_lit l -> doc_lit l
+  | E_id id ->
+      (* TODO replace by a translating via a binding map *)
+      wrap_with_pure as_monadic (string (string_of_id id))
+  | E_lit l -> wrap_with_pure as_monadic (doc_lit l)
+  | E_app (Id_aux (Id "undefined_int", _), _) (* TODO remove when we handle imports *)
+  | E_app (Id_aux (Id "undefined_bit", _), _) (* TODO remove when we handle imports *)
+  | E_app (Id_aux (Id "undefined_bitvector", _), _) (* TODO remove when we handle imports *)
   | E_app (Id_aux (Id "internal_pick", _), _) ->
-      string "sorry" (* TODO replace by actual implementation of internal_pick *)
+      (* TODO replace by actual implementation of internal_pick *)
+      string "sorry"
   | E_internal_plet _ -> string "sorry" (* TODO replace by actual implementation of internal_plet *)
   | E_app (f, args) ->
       let d_id =
         if Env.is_extern f env "lean" then string (Env.get_extern f env "lean")
-        else doc_exp ctx (E_aux (E_id f, (l, annot)))
+        else doc_exp false ctx (E_aux (E_id f, (l, annot)))
       in
-      let d_args = List.map (doc_exp ctx) args in
-      nest 2 (parens (flow (break 1) (d_id :: d_args)))
+      let d_args = List.map d_of_arg args in
+      let fn_monadic = not (Effects.function_is_pure f ctx.global.effect_info) in
+      nest 2 (wrap_with_pure (as_monadic && fn_monadic) (parens (flow (break 1) (d_id :: d_args))))
   | E_vector vals -> failwith "vector found"
-  | E_typ (typ, e) -> parens (separate space [doc_exp ctx e; colon; doc_typ ctx typ])
-  | E_tuple es -> parens (separate_map (comma ^^ space) (doc_exp ctx) es)
+  | E_typ (typ, e) -> wrap_with_pure as_monadic (parens (separate space [doc_exp false ctx e; colon; doc_typ ctx typ]))
```

Comment:
What I was thinking of here was putting the monad's type constructor around the `doc_typ ...`.

Reply:
I don't get it. If `f x : SailM Int`, we want to print this as `(f x : SailM Int)` if `as_monadic` is false and as `pure (f x : SailM Int)` if it is true, right?